tom/changelog.py

```diff
@@ -0,0 +1,153 @@
+            proc = subprocess.run(cmd,
+                    stdout=subprocess.PIPE,
+                    universal_newlines=True,
+                    check=True)
+        finally:
+            os.chdir(olddir)
+        return proc.stdout
+
+    def generate_changelog_in_repo(self, branch, repo):
+        name = repo.repo_name
+        changelog_filename = self.changelog_filename(name)
+        if not changelog_filename:
+            # no need to generate changelog in _this_ repo
+            return False
+        # find old version
+        (header, old_version, old_changelog) = self.split_changelog_into_parts(changelog_filename, repo)
+        new_version = self.get_next_version(old_version, branch)
+        log.debug('next version: {} => {}'.format(old_version, new_version))
+        # checkout new branch
+        timestamp = re.sub('[^0-9-]', '_', str(datetime.datetime.today()))
+        new_branchname = '{}-changelog-{}'.format(new_version, timestamp)
+        repo.checkout(new_branchname, True)
+        if header != '':
+            header += '\n'
+        if name != 'enterprise':
+            # generate changelog only for this repo
+            new_changelog = self.get_changelog_for(name, '--repo', old_version, branch)
+            log.debug('new changelog: [{}]'.format(new_changelog))
+            changelog_contents = header+new_version+':\n'+new_changelog+'\n'+old_changelog+'\n'
+        else:
+            changelog_enterprise = self.get_changelog_for(name, '--enterprise', old_version, branch)
+            log.debug('enterprise changelog:'+changelog_enterprise)
+            changelog_buildscripts = self.get_changelog_for('buildscripts', '--repo', old_version, branch)
+            log.debug('buildscripts changelog:'+changelog_buildscripts)
+            changelog_contents = (new_version+':\n'+changelog_enterprise+'\n'+
+                    '        Packaging changes:\n'+changelog_buildscripts+'\n'+old_changelog+'\n')
+        repo.put_file(changelog_filename, changelog_contents)
+        repo.commit('Added changelog for '+new_version)
+        repo.push(new_branchname)
+        pr_text = self.github.create_pr(
+            'cfengine/'+name, branch, self.username, new_branchname,
+            'Changelog for ' + branch, '')
+        return pr_text
+
+    def run(self, branch):
+        """Generate changelogs on a branch, creating PR in the end"""
+        self.slack.reply("Generating changelogs on " + branch)
+        repo_names = ['core', 'masterfiles', 'enterprise', 'nova', 'mission-portal', 'buildscripts']
+        # checkout all repos to the required branch
+        repos = [GitRepo(
+                    os.path.join(self.repos_root, name),
+                    name,
+                    'cfengine',
+                    self.username,
+                    branch)
+                for name in repo_names]
+        # generate changelogs
+        prs = [self.generate_changelog_in_repo(branch, repo)
+                for repo in repos]
```

Comment:
Wrong indentation, I think.

Reply:
no, I think it's correct (at least according to vim rules :-) ).
It's like function arguments, which are normally written in one line, but when wrapped - indented with 8 spaces:
```
a = func(arg, arg, arg)
a = func(
        arg,
        arg,
        arg)
```

it's a list comprehension, which normally is written in one line, but when wrapped - should be indented with 8 spaces, too, I think:
```
a = [b(c) for c in d]
a = [b(c)
        for c
        in d]
```

Or do you think otherwise?

Reply:
I'd expect the `f` be right below `s`.

Reply:
but that would be _alignment_ instead of _indentation_...

Reply:
Alignment is proper indentation of a sequence of lines. ;)

Reply:
okay...